### ISSUE TITLE
Magicspells version 3.5+ not compatible with latest version of cncp

### ISSUE THREAD

**Issue by Aceakita** [NoCheatPlus/Issues#121]
As soon as i updated MagicSpells to the latest dev build: "MagicSpells-3.5-dev-1603090833" Spells suddenly wasn't being bypassed by NoCheatPlus.

**Comment by asofold** [comment]
Could you post the output of the 'ncp version' command or of 'cncp' ?

I think i had removed MagicSpells support some time ago, because it wasn't provided over maven. Did you configure cncp to exempt from certain instant spells?

**Comment by Aceakita** [comment]
I don't think i have ever changed the configs for either ncp or cncp. Here is the output: http://pastebin.com/nyMAuFfW

As it is right now, i have MagicSpells working perfectly with cncp. All spells works perfectly. But if i update it to 3.5 then cncp will just deny all spells. Thinking their hacks.

**Comment by LaughNgamez** [comment]
Same issue here.

**Comment by LaughNgamez** [comment]
@Aceakita did you ever manage to solve this?

**Comment by Aceakita** [comment]
Nope, i just reverted back to the old version. Since it was working.

**Comment by TheComputerGeek2** [comment]
Well, is there anything that can be done from the MagicSpells side to provide compatibility with no cheat plus?

**Comment by Aceakita** [comment]
Not that i know of. If anyone ever finds a fix for this. Do tell.

**Comment by TheComputerGeek2** [comment]
I was actually hoping to hear from one of the ncp devs on how I can add ncp compatibility to magicspells as I have  been maintaining an alternate version of it.

**Comment by MyPictures** [comment]
@TheComputerGeek2 Compatibility would be done as in every plugin over the exemption API. Example: https://github.com/asofold/CompatNoCheatPlus/blob/master/CompatNoCheatPlus/src/me/asofold/bpl/cncp/hooks/magicspells/HookMagicSpells.java

Sorry but I cannot help you much in this subject. API and cncp is handled by @asofold fully.

**Comment by LaughNgamez** [comment]
@asofold have anything to say about this? Still hoping for a fix.

**Comment by asofold** [comment]
Perhaps i can bring it back, or we contact the developer. We just had a data loss issue, so there has been some delay last week.

**Comment by LaughNgamez** [comment]
@asofold  Thanks for the response!

Currently the main developer Nisovin (who I assumed worked out the latest NCP compatibility) is likely out of reach. @TheComputerGeek2 maintains a well known fork of the project and is interested in fixing it from what I know.

**Comment by asofold** [comment]
@LaughNgamez That would be very good. For instant/loop-like effects, simple temporary exemption will do (metadata is possible, or access the NCP API). Is that their GitHub account name?

**Comment by TheComputerGeek2** [comment]
@asofold  I'd happily help, what can I do to make this process easier? (I'm the guy maintaining it, by the way)

**Comment by asofold** [comment]
@TheComputerGeek2 

This will depend on what kind of methods we can actually use.

There are different cases how events fire, some confining the possibilities.

1. Implicitly firing events, such as a standard Melee attack with calling damage including a damaging player, or a standard block break event following upon BlockDamageEvent.setInstaBreak.
2. Events fully under your control, which you can consistently derive/name. E.g. SuperHammerExtraBlockBreakEvent. For these we could extend NCP by adding configuration for generic exemption triggers by event classes.
3. Things that somehow affect gameplay in a way currently incompatible with NCP, such as sequences of setting velocity and hunger or switching attributes/speeds/effects and similar frequently on server side, usually without having an event fire for them. While we want to support all/most (default :p -) vanilla behavior, this can be quite testing due to latency issues.
 
For 2. we usually can 'somehow do something' on either side. For 1. the only thing that usually works is wrapping the entire thing with exemption. Preferably that's done within the other plugin, but we have done so for mcMMO in the past within CompatNoCheatPlus. In general it's much better to have the skill-plugin cause temporary exemption, because it'll  usually be pinpointed.

To start with something clearly doable i suggest tackling all 'instant' effects concerning damage and block break/place and such. These can be done with temporary exemption, for which we currently provide two ways:
a) Exemption by certain metadata (add before and remove after loop/thing). This is very coarse and wild-card exempts from all checks at present, but it could be refined if necessary. Slightly risky would be, if you forget to or can't remove the metadata, due to bugs or special circumstances, resulting in wild-card exemption.
b) Exemption by hooking into the NCP API. This isn't very complex by itself, but involves plugin dependencies and managing if to use such a hook at all. Possibly less convenient.

(_There could be another way, which might be feasible in case of multiple events per incident, such as area effects. That would involve firing a custom event  before applying the effects and another custom event after being done (possibly the same event, having a boolean method for start vs. end and an integer for incident counter). This could allow generic pinpointing with other than metadata, using some reflection though. Possibly this doesn't even perform better than using metada would. Just to note another way of wrapping with exemption._)

The generic exemption without need to hook into a specific anti cheating plugin is a current line of thought, not really an established thing.

**Comment by TheComputerGeek2** [comment]
@asofold So what I'm currently working on doing is changing all regular spigot events fired by magicspells, that are likely already being checked by ncp, to subclasses of the respective event but having all of those implement an empty interface to group them all together. The idea of this is that you will be able to just check to see if the event is an instance of said interface and keeping things (hopefully) simpler on your end. I can provide an example of this if you want.

Although there is also something that I would like a little clarification on. And that is what conditions do events need to be fired. For example, one of my users has been having problems with the leap spell but I can't seem to identify what event ncp would be triggering on as I can't find one manually being fired there. My only remaining guess is that the problem is a lack of event indicating the velocity change validity. Do you know of any way to tell ncp that this change is acceptable through firing events?

Also, just something to note: I am trying to avoid using time span exemptions for anything since for something such as a pulser spell (which may hit multiple targets with damage and potentially at a range), granting time period exemption would be unsafe. The exploit would be to initiate the spell to gain exemption, and then enable client mods while the spell is active. I'm sorry if this is a bit too wordy and I can try rephrasing it if need be, but I wanted to express my goals.

**Comment by asofold** [comment]
Concerning events: Such will allow much easier compatibility hooks for wrapping things. _Currently the situation is more like having an over-load with vehicle-fly-and-possibly-more-related~thing, so i'll likely discontinue CompatNoCheatPlus as is. NCP won't hook directly into plugins, but might instead use reflection or similar to check for class names and interfaces, with the advantage of having that part built into the core plugin and having it configurable. Might change my mind on particularly easy to overview and maintain aspects, but essentially those things are better inside of NCP, though i won't want to have NCP build with dependencies that are not absolutely necessary for the core functionality._

Concerning firing events: If you set velocity, this is intended to be supported out of the box by NCP, so there is a strong intent to fix those issues "simply". So there it would be best to produce debug logs with the specific feature leading to a false positive with NCP, best have several instances of the issue in a log. Setting the velocity leads to a velocity event, so NCP should notice, but sometimes there are other side conditions like hunger, or just bugs like resetting internal velocity accounting for some reason.

Timed exemptions are the worst and should be avoided :), or at least confined to the expected check-alerts and envelope as much as possible. That's why i said we should first focus on the 'instantly' applied effects and/or wrapping loops/events with exemption.

One example for exemption without hooking into NCP: If you have a loop firing multiple secondary area damage events (directly or indirectly by entity.damage(123, player)), then you could add the  metadata 'nocheat.exempt' to the player before starting the loop (preferable fixed boolean -> true) and remove it after the loop. This would serve like temporary exemption from all NCP checks, just don't forget to remove the metadata after applying some effect. This is already possible (wildcard exemption by metadata 'nocheat.exempt'), NCP just tests for hasMetadata(...), though i'd suggest using a fixed boolean value there. This part could be elaborated to contain some structure, if more pinpointing or "future safeguard" is intended (e.g. apply once per event only), but currently the simple thing is there to see how others think about the metadata approach.

**Comment by TheComputerGeek2** [comment]
Is this metadata approach something that ncp already checks for? Or are you saying this is something that it could do in the future?

**Comment by asofold** [comment]
@TheComputerGeek2 I am not sure about performance for meta data, i assume this'll be hard to measure, unless you can compile the support and use away and compare timings for the same (big) server.

With metadata NCP currently only supports wildcard exemption for all checks, there is no parsing of values done yet. This would be something to discuss, e.g. one could use a boolean value for wildcard or a List<String> thing, with a defined format, e.g. check types to exempt from, or even advanced declarations. E.g. the list could have some fixedid:....uuid entry and always use the exact same strings, so plugins can store the effects of that very instance, syntax could be extended to cover expiration, special use. I'd also allow addressing checks in a plugin-independent way, as well as addressing a certain plugin. So this is much open to discussion, could also imagine providing some abstract base implementation somehow (for both sides), say under LGPL. I just couldn't find a directly satisfying syntax concept back then.

**Comment by TheComputerGeek2** [comment]
@asofold If you don't mind, I'd actually like to try my hand at making a suitable library for exemptions, I have an idea on how to set it up so that plugins don't conflict with each other and how to deal with multiple exemption requests with partially overlapping time spans. I'll push my ideas to a repository soon.

**Comment by asofold** [comment]
Sure, also feel free to discuss details. It just needs to do without plugin dependencies and have reasonable performance.

Possibly interesting topics:
* Performance
 * Re-use keys (and values).
 * Re-use strings whenever possible.
 * Provide short-cuts to skip parsing/processing entire values. E.g. the first entry of a value may contain something like (at)shortcut:(uu/id).
 * Support storing processed values / short-cut content by design.
 * All lower case, also plugin names. Do without any string processing except when encountering a token/value first time.
* Allow to address generic areas/checks (plugin independent).
 * Not entirely sure, think of some entry like generic.damage.area, to let plugins decide what to exempt from.
* Allow addressing certain plugins.
 * E.g. an entry like nocheatplus:fight.targeting.
* Thinkable complex syntax, to be added to an exemption context.
 * Indicate future events (+ number of?) to exempt from. (Not sure, but BlockDamageEvent.setInstaBreak will lead to another BlockBreakEvent firing - though that should be supported out of the box, there may be other similar cases.)
 * Indicate validity (this event only, max. remove next tick (bad-ish), )
* Complex definitions to avoid plugins overriding each others entries.
 * Worse performance for sure :p.
 * Possibly something you might want to be in. 
  * You could also have one string per plugin within a list of exemptions, but since values are tied to plugins, this seems bad.
  * You could have a generic lower case key bound to plugin name - tied to activated plugins only. This way anti cheating plugins can easily keep track of what to check for at all, by a combination of configuration (which plugins might do something at all) and monitoring plugins enabling. Possibly plugins supporting this could just implement an interface of a certain name in order to indicate that they use exemption by metadata, thus an anti cheating plugin can the relay to checking for the generic plugin exemption key (e.g. exemptions:pluginname -> value spec). An extension could be that the plugin specifies, what type of exemptions exist at all, which then will allow better performance optimization on side of other plugins [hardly anyone might implement that].

 Just to touch my direction of thought with text.

**Comment by TheComputerGeek2** [comment]
At the moment, what I'm designing is more specific to NoCheatPlus with the goal of being able to have safe, non-conflicting, exemptions. I plan to expand it to a more general base afterwards. This would serve as a way for plugins to easily hook into NoCheatPlus exemption without worrying about how to ensure that their exemptions do not conflict with others. This structure already will allow for a lot of optimization room on the plugins hooking exemption as they will control the ticket management.
The way I see it, the thing that people will benefit most from is having safe exemptions from ncp as it is a very common plugin and there isn't really much of a stable exemption system that I know of which doesn't suffer from exemption conflict. After that is taken care of, we can refocus on the more generalized solution. And given the current topic of the issue, I think that people would agree in that regard.

As a side note: I figured out what was going on with leap. Nisovin had added a field called client-only and defaulted it to true. I didn't realize that until somewhat recently. So I resolved that on my end by setting it to false by default. Although I still want to produce proper compatibility for NCP and magicspells.

**Comment by asofold** [comment]
@TheComputerGeek2 The leap thing sounds great :), but wait a second! Need to ask, in order to clarify: If you're mostly interested to have NCP have safe exemption for multiple plugins potentially interfering, shouldn't that be built into NCP directly? That only applies for using the NCP exemption API then, though - it's actually kind of planned, to have the possibility to add exemptions that are addressable by some kind of id, and to have light/temporary exemptions that are removed on any call to unexempt(player/thing).

If you stay with heading for a way to get exemption via metadata or something else that doesn't need plugin dependencies, then certainly go ahead at any speed :).

**Comment by TheComputerGeek2** [comment]
@asofold the idea with what I'm working on is that it hopefully will become polished enough to be migrated into NCP. The idea of it is that instead of saying exempt(player, check) and then later unexempt(player, check) you create an object containing the checks that you want to be bypassed. you then say exempt(player, ticket) and unexempt(player, ticket). Multiple tickets can cover the same checks and as long as at least one active ticket is exempting a check, that check will be exempted for that player. The idea with this is that you can have multiple plugins that request exemption and never have to worry about a condition where it either unexempts too early due to another plugin unexempting after completing its own process, or the condition of never exempting due to checks of it already being exempt at the beginning. Tickets can be reused as needed. Does this make any sense?

The client-only switch was where instead of actually setting the player velocity server side, it just sent packets to the client telling the client that the velocity had changed.

**Comment by asofold** [comment]
@TheComputerGeek2 The ticket approach looks very good.

For simplicity of use, at some point i'd also allow a quick thing with implicit tickets, like long id = exempt(thisnthat); dosomething(); unexempt(id).

On ncp side, the check types will be changed to classes at some point, but they will be keeping the tree-like structure. Future plan was to switch from expanding an exemption to all child check types to just walking along the line of parents to see if an exemption is present. Not sure that kind of information is relevant, but there'll be some refadctoring / changes of internals later on, the exemption API will not change by that much though (from tickets on), other than possibly moving it from static to less/non-static.

**Comment by TheComputerGeek2** [comment]
@asofold the idea of using the explicit tickets is that it would (hopefully) be intuitive for people to hook into and that it would allow for the automatic removal of exemptions when a ticket is no longer held by the plugin that requested it. In addition, it would make ticket reuse much easier. This would easily be done through the use of a weak hash set (just the set view of a weak hash map). I figure that it is better for a plugin that is hooking into ncp to fail to exempt than to fail to unexempt.
Although I can see implicit tickets still happening for the sake of simplicity. of exemptions.

**Comment by asofold** [comment]
@TheComputerGeek2 How are things going?

Had a look at this: https://github.com/TheComputerGeek2/NCPExemptions

I'll some point soon consider adding something similar to NCP, to support that ticket-like setup directly. For metadata based tickets there may be other future additions to translate between meta-data-based String tickets and the NCP internals.

Some things ill consider to add to tickets:
* Expiration handling ~ might add information, if to expire on logout, plugin disable, if to tie to a plugin at all, if to remove exemptions on wildcard unexemption. Such...
* Overriding properties - what with setting the same ticket multiple times (just needs a concept, possibly just ignore).
* Possibly add a way to construct tickets using another ticket or basic settings as default (ticket.duplicateForCheckTypes(....)).
* Likely store by uuid, to be able to keep exemptions over login.
* Possibly rename methods (addExemptionTicket, removeExemptionTicket?).

I could try to bring this in with a major refactoring towards non-enum check types. A dynamic check type registry cares for the tree structure and possibly legacy mappings and all, alongside with that there may then be other internals implemented for traversing exemption-nodes.

**Comment by TheComputerGeek2** [comment]
@asofold would you like me to see about implementing these on my NCPExemptions repository? If you want, I could see about pushing it to maven (once I figure out how, haven't done much with maven yet) then you could just shade it into NoCheatPlus and use it for exemption checking along side where you check for the "nocheatplus.exempt" on player metadata.

**Comment by asofold** [comment]
@TheComputerGeek2 For internals of NCP i would have to redo lots anyway, most likely there'll be a tree-like structure per player with exemptions and there might be some type of nodes implemented that can't really be identified with the tickets - also the check types will change medium term. So if the implementation/sketch is meant to provide a meta registry, there might be some kind of adapter necessary, after functionality has really been implemented on side of NCP, because the internals are meant to be very efficient and fast. The approach you suggest now would be that NCP asks at that service if there is a ticket and then applies exemption accordingly? Even then i'd want the internals to be optimized (map external tickets fastest way to internal ones for reuse), possibly lay out information across nodes on first registration.

Actually i may have had generic cross plugin compatibiltiy in mind when reading the code, but you meant sketching out an approach to use with NCP "dedicated" at first? For generic cross-plugin compatibility one would best do without depending on external libraries or plugins. Helper classes could then generate metadata or whichever tickets only using basic objects/strings or using the service registry, though the generic service registry isn't very nice to use (could think of a reference implementation that helps setting up generic tickets, accessing the registry and get + encapsulate the appropriate services neatly, so the calls to exempt/unexempt only use very simple registry objects or Callable<simple> stuff, that can be efficiently used/reused at runtime, only setup would use reflection).

Sorry, bit in a rush, so don't rush!

**Comment by TheComputerGeek2** [comment]
@asofold 
Just remembered this thread, and as for Magicspells, I think that, at this point, you could just add exemptions based upon what class the events come from, since (I'm pretty sure) all Magicspells thrown versions of game activity events such as entity damage by entity, are thrown as a magicspells specified subclass of the event. To make things even easier, all of these events implement an empty interface called IMagicSpellsCompatEvent. You can take a look at the events here https://github.com/TheComputerGeek2/MagicSpells/tree/master/src/com/nisovin/magicspells/events

Is there anything else I would need to do for an event class based exemption to be possible?

**Comment by asofold** [comment]
The PR looks good - i haven't tested it though :). I'll try to put that generic stuff into NCP as soon as there is a little more free time/space.

A common interface will make it possible to auto-suggest event names later on, possibly i could add configuring event-class based exemption from in-game/console some day, thinking of custom effects. Spell id's or names would then allow relating to them.

**Comment by TheComputerGeek2** [comment]
Can anyone still reproduce the bug on MagicSpells 3.5 (my distro, sorry, won't work with nisovin's) after updating ncpc? @Aceakita @LaughNgamez 

@asofold the best way to identify a spell's operations is by the spell class. If it would help, I could open up more of the spell fields to allow ncpc to read it more easily without the use of reflection. And what did you mean by custom effects? I assume you are not referring to particle shows.

**Comment by asofold** [comment]
Concerning exemption as is, one should expect it to be 'ok' to unexempt on monitor priority for an event (preferably of the most outer event in  case of nested events). The current implementation of exemption is not really useful for longer term exemption in a production environment.

Guidelines... yeah. I wanted to get rid of cncp before creating guidelines though (put reflection based stuff into NCP then).

**Comment by asofold** [comment]
Assume fixed for now - unfortunately can't extend any exemption based stuff until PlayerData is used as the definitive per-player data storage (won't touch the static utility internally anymore).
